File: src/messaging/dotnet/src/MessageRouterAdapter/MessageRouterMessaging.cs
// Morgan Stanley makes this available to you under the Apache License,
// Version 2.0 (the "License"). You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0.
//
// See the NOTICE file distributed with this work for additional information
// regarding copyright ownership. Unless required by applicable law or agreed
// to in writing, software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
// or implied. See the License for the specific language governing permissions
// and limitations under the License.

using MorganStanley.ComposeUI.Messaging;
using MorganStanley.ComposeUI.Messaging.Abstractions;
using MorganStanley.ComposeUI.Messaging.Abstractions.Exceptions;
using MorganStanley.ComposeUI.MessagingAdapter.MessageRouter;

using TopicMessageHandler = MorganStanley.ComposeUI.Messaging.Abstractions.TopicMessageHandler;

namespace MorganStanley.ComposeUI.MessagingAdapter;

/// <summary>
/// Provides an implementation of <see cref="IMessaging"/> that wraps an <see cref="IMessagingService"/> instance,
/// offering methods for connecting, invoking, publishing, subscribing, and service registration with consistent exception handling.
/// </summary>
/// <remarks>
/// This class adapts the lower-level <see cref="IMessagingService"/> interface to the <see cref="IMessaging"/> abstraction,
/// and ensures that exceptions from the message router are wrapped in adapter-specific exceptions.
/// </remarks>
internal class MessageRouterMessaging : IMessaging
{
    private readonly IMessageRouter _messageRouter;
    public string? ClientId => _messageRouter.ClientId;

    /// <summary>
    /// Initializes a new instance of the <see cref="MessageRouterMessaging"/> class with the specified messaging service.
    /// </summary>
    /// <param name="messageRouter">The underlying <see cref="IMessagingService"/> to be wrapped by this adapter.</param>
    public MessageRouterMessaging(IMessageRouter messageRouter)
    {
        _messageRouter = messageRouter;
    }

    /// <summary>
    /// Subscribes to the specified topic and invokes the subscriber callback when a message is received.
    /// </summary>
    /// <param name="topic">The topic to subscribe to.</param>
    /// <param name="subscriber">The callback to invoke when a message is received.</param>
    /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
    /// <returns>
    /// A <see cref="ValueTask{IDisposable}"/> representing the asynchronous subscription operation.
    /// The result contains an <see cref="IDisposable"/> that can be used to unsubscribe.
    /// </returns>
    public async ValueTask<IAsyncDisposable> SubscribeAsync(
        string topic,
        TopicMessageHandler subscriber,
        CancellationToken cancellationToken = default)
    {
        return await WrapMessageRouterExceptionsAsync(async () =>
        {
            var asyncSubscriber = new MessageHandler(buffer =>
            {
                var message = buffer.GetString();
                return subscriber(message);
            });
            var asyncDisposable = await _messageRouter.SubscribeAsync(topic, asyncSubscriber, cancellationToken);

            return asyncDisposable;
        });
    }

    /// <summary>
    /// Publishes a message to the specified topic.
    /// </summary>
    /// <param name="topic">The topic to publish to.</param>
    /// <param name="message">The message to publish, or null.</param>
    /// <param name="options">Optional publish options.</param>
    /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
    /// <returns>A <see cref="ValueTask"/> representing the asynchronous operation.</returns>
    public ValueTask PublishAsync(
        string topic,
        string? message = null,
        CancellationToken cancellationToken = default) => WrapMessageRouterExceptionsAsync(() =>
        _messageRouter.PublishAsync(
            topic,
            MessageBuffer.Create(message ?? "null"),
            default,
            cancellationToken));

    /// <summary>
    /// Registers a service endpoint with a subscriber callback.
    /// </summary>
    /// <param name="endpoint">The endpoint to register.</param>
    /// <param name="subscriber">The callback to invoke when a message is received.</param>
    /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
    /// <returns>A <see cref="ValueTask"/> representing the asynchronous operation.</returns>
    public async ValueTask<IAsyncDisposable> RegisterServiceAsync(string serviceName, Messaging.Abstractions.ServiceHandler subscriber, CancellationToken cancellationToken = default)
    {
        await WrapMessageRouterExceptionsAsync(() =>
        _messageRouter.RegisterServiceAsync(serviceName, async (name, buffer, context) =>
        {
            var payload = buffer?.GetString() ?? string.Empty;
            var result = await subscriber(payload);

            return string.IsNullOrEmpty(result) ? null : MessageBuffer.Create(result);
        }, cancellationToken));
        return new MessageRouterServiceRegistration(serviceName, _messageRouter);
    }


    /// <summary>
    /// Invokes an endpoint asynchronously and returns the response as a string.
    /// </summary>
    /// <param name="endpoint">The endpoint to invoke.</param>
    /// <param name="payload">The payload to send, or null.</param>
    /// <param name="options">Optional invoke options.</param>
    /// <param name="cancellationToken">A cancellation token to observe while waiting for the task to complete.</param>
    /// <returns>A <see cref="ValueTask{TResult}"/> representing the asynchronous operation, with the response as a string.</returns>
    public ValueTask<string?> InvokeServiceAsync(
        string serviceName,
        string? payload = null,
        CancellationToken cancellationToken = default) => WrapMessageRouterExceptionsAsync(async () =>
        {
            var responseBuffer = await WrapMessageRouterExceptionsAsync(() => _messageRouter.InvokeAsync(
                serviceName,
                MessageBuffer.Create(payload ?? "null"),
                default,
                cancellationToken));
            return responseBuffer?.GetString();
        });

    /// <summary>
    /// Executes the specified asynchronous operation and wraps MessageRouter exceptions with MessagingAdapter exceptions.
    /// </summary>
    /// <param name="operation">The asynchronous operation to execute.</param>
    /// <returns>A ValueTask representing the asynchronous operation.</returns>
    private static async ValueTask WrapMessageRouterExceptionsAsync(Func<ValueTask> operation)
    {
        try
        {
            await operation();
        }
        catch (MessageRouterDuplicateEndpointException ex)
        {
            throw new DuplicateServiceNameException(ex.Name, ex.Message, ex);
        }
        catch (MessageRouterException ex)
        {
            throw new MessagingException(ex.Name, ex.Message, ex);
        }
    }

    /// <summary>
    /// Executes the specified asynchronous operation and wraps MessageRouter exceptions with MessagingAdapter exceptions.
    /// </summary>
    /// <typeparam name="TResult">The result type of the asynchronous operation.</typeparam>
    /// <param name="operation">The asynchronous operation to execute.</param>
    /// <returns>A ValueTask representing the asynchronous operation with a result.</returns>
    private static async ValueTask<TResult> WrapMessageRouterExceptionsAsync<TResult>(Func<ValueTask<TResult>> operation)
    {
        try
        {
            return await operation();
        }
        catch (MessageRouterDuplicateEndpointException ex)
        {
            throw new DuplicateServiceNameException(ex.Name, ex.Message, ex);
        }
        catch (MessageRouterException ex)
        {
            throw new MessagingException(ex.Name, ex.Message, ex);
        }
    }
}

File: src/messaging/js/composeui-messaging-message-router/src/MessageRouterMessaging.ts
/* 
 *  Morgan Stanley makes this available to you under the Apache License,
 *  Version 2.0 (the "License"). You may obtain a copy of the License at
 *       http://www.apache.org/licenses/LICENSE-2.0.
 *  See the NOTICE file distributed with this work for additional information
 *  regarding copyright ownership. Unless required by applicable law or agreed
 *  to in writing, software distributed under the License is distributed on an
 *  "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 *  or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 *  
 */

import { IMessaging, ServiceHandler, TopicMessageHandler } from "@morgan-stanley/composeui-messaging-abstractions";
import { MessageRouter } from "@morgan-stanley/composeui-messaging-client";
import { Unsubscribable } from "rxjs";
import { AsyncDisposableWrapper } from "./AsyncDisposableWrapper";

/**
 * Implementation of IMessaging interface using MessageRouter.
 * Provides messaging capabilities through the MessageRouter client for ComposeUI applications.
 */
export class MessageRouterMessaging implements IMessaging {
    /**
     * Creates a new instance of MessageRouterMessaging.
     * @param messageRouterClient The MessageRouter client instance to use for communication.
     */
    constructor(private readonly messageRouterClient: MessageRouter) {
    }

    /**
     * Subscribes to messages on a specific topic.
     * @param topic The topic to subscribe to.
     * @param subscriber Callback function that will be invoked with each received message.
     * @param cancellationToken Optional signal to cancel the subscription setup.
     * @returns A Promise that resolves to an Unsubscribable object for managing the subscription.
     * @remarks If a message is received without a payload, a warning will be logged and the subscriber will not be called.
     */
    public subscribe(topic: string, subscriber: TopicMessageHandler, cancellationToken?: AbortSignal): Promise<Unsubscribable> {

        return this.messageRouterClient.subscribe(topic, (message) => {
            if (!message.payload) {
                console.warn(`Received empty message on topic ${topic}`);
                return;
            }
            subscriber(message.payload);
        });
    }

    /**
     * Publishes a message to a specific topic.
     * @param topic The topic to publish to.
     * @param message The message content to publish.
     * @param cancellationToken Optional signal to cancel the publish operation.
     * @returns A Promise that resolves when the message has been published.
     */
    public publish(topic: string, message: string, cancellationToken?: AbortSignal): Promise<void> {
        return this.messageRouterClient.publish(topic, message);
    }

    /**
     * Registers a service handler for a specific service name.
     * @param serviceName The name of the service to register.
     * @param serviceHandler The handler function that will process service requests.
     * @param cancellationToken Optional signal to cancel the service registration.
     * @returns A Promise that resolves to an AsyncDisposable for managing the service registration.
     * @remarks The service handler will receive the payload from the request and should return a response.
     * Both the payload and response can be null.
     */
    public async registerService(serviceName: string, serviceHandler: ServiceHandler, cancellationToken?: AbortSignal): Promise<AsyncDisposable> {
        await this.messageRouterClient.registerService(serviceName, async (endpoint, payload, context) => {
             const result = await serviceHandler(payload!);
                return result!;
        });
        
        const disposable = new AsyncDisposableWrapper(this.messageRouterClient, serviceName);
        return disposable;
    }

    /**
     * Invokes a registered service.
     * @param serviceName The name of the service to invoke.
     * @param payload Optional payload to send with the service request.
     * @param cancellationToken Optional signal to cancel the service invocation.
     * @returns A Promise that resolves to the service response or null if no response is received.
     * @remarks If the payload is null, the service will be invoked without a payload.
     * The response will be null if the service doesn't return a response or if an error occurs.
     */
    public async invokeService(serviceName: string, payload?: string | null, cancellationToken?: AbortSignal): Promise<string | null> {
        if (payload == null) {
            const result =  await this.messageRouterClient.invoke(serviceName);
            if (!result){
                return null;
            }

            return result;
        }

        const response = await this.messageRouterClient.invoke(serviceName, payload);
        if (!response) {
            return null;
        }
        return response;
    }
}